src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs

```diff
@@ -83,12 +93,14 @@ internal string HostId
 
         internal ConcurrentDictionary<string, DiagnosticEvent> Events => _events;
 
+        internal IPrimaryHostStateProvider HostStateProvider => _primaryHostStateProvider ??= _serviceProvider?.GetService<IPrimaryHostStateProvider>();
```

Comment:
I don't think we can cache this... because of the scopes potentially being different with WebHost vs JobHost, you could end up holding onto an `IPrimaryHostStateProvider` for a host that has been disposed.

I think that every time we need to run through this and take whatever `IPrimaryHostStateProvder` is returned... or skip the purge if null, which you're already doing below. Perf isn't really a concern here so it shouldn't be a big deal to not cache this.

Would recommend making this property into something like `GetPrimaryHostStateProvider()` that tries to resolve the service every time.

Reply:
At yup good catch, was not intending to cache this. No coding after 10pm for me :)